libs/test-utils/src/mock_function.ts

```diff
@@ -0,0 +1,180 @@
+import { inspect } from 'util';
+import deepEqual from 'deep-eql';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AnyFunc = (...args: any[]) => any;
+
+interface Call<Func extends AnyFunc> {
+  input: Parameters<Func>;
+}
+
+type ExpectedCall<Func extends AnyFunc> =
+  | ({ output: ReturnType<Func> } & Call<Func>)
+  | ({ error: unknown } & Call<Func>);
+
+interface ReturnHelpers<Func extends AnyFunc> {
+  returns(output: ReturnType<Func>): void;
+  throws(error: unknown): void;
+}
+
+interface AsyncReturnHelpers<Func extends AnyFunc> {
+  resolves(output: Awaited<ReturnType<Func>>): void;
+  throws(error: unknown): void;
+}
+
+export interface MockFunction<Func extends AnyFunc> {
+  (...input: Parameters<Func>): ReturnType<Func>;
+  expectCallWith(
+    ...input: Parameters<Func>
+  ): ReturnType<Func> extends Promise<unknown>
+    ? AsyncReturnHelpers<Func>
+    : ReturnHelpers<Func>;
+  reset(): void;
+  assertComplete(): void;
+}
+
+interface MockFunctionState<Func extends AnyFunc> {
+  expectedCalls: Array<ExpectedCall<Func>>;
+  actualCalls: Array<Call<Func>>;
+}
+
+function formatFunctionCall(name: string, input: unknown[]): string {
+  return `${name}(${input
+    .map((arg) => inspect(arg, { depth: null }))
+    .join(', ')})`;
+}
+
+/**
+ * Creates a mock function, similar to jest.fn(), but with stricter rules to
+ * help you avoid making unintentional mistakes.
+ *
+ * Example usage:
+ *
+ *  // Let's say we have a real function, add, that we want to mock
+ *  function add(num1: number, num2: number): number {
+ *    return num1 + num2;
+ *  }
+ *
+ *  // We can create a mock function that behaves the same way
+ *  const addMock = mockFunction<typeof add>('add');
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock(1, 2); // -> returns 3
+ *  addMock.expectCallWith(1, NaN).throws(new Error('Cant add NaN'));
+ *  addMock(1, NaN); // -> throws Error('Cant add NaN')
+ *
+ *  // Unlike jest.fn(), mockFunction is strict...
+ *
+ *  // Each expected call can only be used once, in order
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock(1, 2); // -> returns 3
+ *  addMock(1, 2); // -> throws an error because we didn't expect this call
+ *
+ *  // The actual input must match the expected input
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock(1, 3); // -> throws an error because the input doesn't match
+ *
+ *  // Every expected call must be used
+ *  addMock.expectCallWith(1, 2).returns(3);
+ *  addMock.assertComplete(); // -> throws an error because we didn't use the expected call
+ *
+ * Recommendations:
+ * - Always call mockFunction.assertComplete() to ensure that all expected calls were used
```

Comment:
Down the road, could be nice to find a way to make it such that writers of tests don't have to remember to do this - I could definitely see someone forgetting to do so. No simple ideas are coming to mind for this

Reply:
Yeah I noodled on this for a while and couldn't think of anything either. The best idea I had was making it so you could only instantiate a mock function by calling a wrapper function like this:

```ts
withMockFunction(mockFunction => ... use the mock function ... )
```

Then the wrapper can run `assertCompleted` logic after the callback finishes. This is the approach used by `withMockFetch` in Arlo.

I opted for not doing this because it doesn't mesh well with jest's philosophy about using `before` and `after` for setup/cleanup. Even though I don't like that philosophy because it encourages shared global state between tests, I was trying to play nice. Definitely willing to reconsider this decision though!

Luckily `assertCompleted` really only matters for catching extra expected mock calls, so if you forget to call it the worst thing that will happen is that you'll have some cruft in your test.

Reply:
Ah yes, that makes the motivation for the wrapper function approach in Arlo clearer!

I also agree that 1) we should try to stay in line with Jest principles and 2) the implications of missing `assertCompleted` aren't that bad, so I'm okay punting on this

Implementing some kind of shared global state to make things more automatic also definitely increases the risk I mentioned in my top-level comment of the handmade alternative becoming harder and harder to maintain

Reply:
Yeah, one thing I actually like a lot about pytest is that they let you run code after the test by using a generator for the fixture factory function. The fixture gets yielded and then after the test runs, control returns to the fixture factory function (example: https://github.com/votingworks/arlo/blob/39a8bf585ddf40558201e7e75779e7d2cbf199e8/server/tests/test_background_tasks.py#L31-L44). I wish there were something like that for jest